providers/azure.go

```diff
@@ -149,13 +145,103 @@ func (p *AzureProvider) Redeem(ctx context.Context, redirectURL, code string) (*
 	created := time.Now()
 	expires := time.Unix(jsonResponse.ExpiresOn, 0)
 
-	return &sessions.SessionState{
+	session := &sessions.SessionState{
 		AccessToken:  jsonResponse.AccessToken,
 		IDToken:      jsonResponse.IDToken,
 		CreatedAt:    &created,
 		ExpiresOn:    &expires,
 		RefreshToken: jsonResponse.RefreshToken,
-	}, nil
+	}
+
+	email, err := p.verifyTokensAndExtractEmail(ctx, session.IDToken, session.AccessToken)
+	if err != nil {
+		return nil, err
+	}
+	session.Email = email
+
+	return session, nil
+}
+
+// EnrichSession finds the email to enrich the session state
+func (p *AzureProvider) EnrichSession(ctx context.Context, s *sessions.SessionState) error {
+	if s.Email != "" {
+		return nil
+	}
+
+	email, err := getEmailFromProfileAPI(ctx, s.AccessToken, p.ProfileURL.String())
+	if email == "" || err != nil {
+		if err != nil {
+			err = fmt.Errorf("unable to get email address, error: %v", err)
+		} else {
+			err = errors.New("unable to get email address")
+		}
+		logger.Errorf("%s", err)
+		return err
+	}
```

Comment:
This nesting is a bit redundant - wen can clean it up the logical layers by returning early
```suggestion
        if err != nil {
	        return fmt.Errorf("unable to get email address: %v", err)
        }
        if email == "" {
	        return errors.New("unable to get email address")
        }
```